### PR TITLE
Fixed Log disappears if network not connected for websocket and TCP-Client

### PR DESCRIPTION
**Pull request by shivam5594**
Features added 
1) Robust websocket connection, logs preserved when server is down and sent in chronologically once it is up. 
2) Even when client dies and comes up again, it will be able to connect to server without any issues. 
3) TCP-Client is now better, can reconnect to server if it goes down and will send logs later in chronological order with least data-loss. Tried at 100Hz data logging. 
4) removed unnecessary logs to avoid over-flowing.

### CODE REVIEW

File: tcp_handler.c [hunk 1]
```diff
@@ -21,22 +21,25 @@ void tcp_network_manager(struct tcp_network_data* nm)
     if (nm->sock < 0)
     {
         ESP_LOGE(TAG, "Unable to create socket: errno %d", errno);
+        tcp_close_network_manager(nm);
```
Comment (VedantParanjape) [inline tcp_handler.c]: Is it right to close network manager, when it isn't running ?

File: websocket_handler.c [hunk 2]
```diff
@@ -66,7 +71,6 @@ esp_websocket_client_handle_t websocket_network_manager()
  */
 int websocket_send_data(esp_websocket_client_handle_t network_handle, char* payload)
```
Comment (VedantParanjape) [inline websocket_handler.c:73]: Can you fix whitespaces in this function

File: include/tcp_handler.h [hunk 3]
```diff
@@ -18,6 +18,10 @@
 #include "lwip/sys.h"
 #include <lwip/netdb.h>
 
+#ifdef __cplusplus
```
Comment (VedantParanjape) [inline include/tcp_handler.h:21]: Why is this needed here?

File: include/udp_handler.h [hunk 4]
```diff
@@ -18,6 +18,10 @@
 #include "lwip/sys.h"
 #include <lwip/netdb.h>
 
+#ifdef __cplusplus
```
Comment (VedantParanjape) [inline include/udp_handler.h:21]: Why is this needed ? I don't want to use this with C++

File: include/tcp_handler.h [hunk 5]
```diff
@@ -31,9 +35,13 @@ struct tcp_network_data
     int sock;
 };
 
-void tcp_network_manager(struct tcp_network_data* nm);
+bool tcp_network_manager(struct tcp_network_data* nm);
```
Comment (VedantParanjape) [inline include/tcp_handler.h:38]: Can you use int rather than bool ?

File: include/wifi_logger.h [hunk 6]
```diff
@@ -27,6 +31,11 @@ void generate_log_message(esp_log_level_t level, const char *TAG, int line, cons
 int system_log_message_route(const char* fmt, va_list tag);
 void start_wifi_logger(void);
 void wifi_logger();
+bool is_connected(void* handle_t);
```
Comment (VedantParanjape) [inline include/wifi_logger.h:34]: Can you use int instead of bool

File: tcp_handler.c (hunk 1 above)
Comment (shivam5594) [reply tcp_handler.c]: In TCP you need to close it entirely to create a new one to reconnect. So whenever it fails, it closes it and recreates socket.

File: websocket_handler.c (hunk 2 above)
Comment (shivam5594) [reply websocket_handler.c:73]: whitespace as in tabs and space?

File: wifi_logger.c [hunk 7]
```diff
@@ -192,11 +192,38 @@ void start_wifi_logger(void)
...
     ESP_LOGI(tag_wifi_logger, "WiFi logger initialised");
 }
 
+/*
+ * @brief A common wrapper function to check connection status for all interfaces
+ *
+ * @param handle_t Interface handle
+ * @return bool True if connected
+ */
+bool is_connected(void* handle_t)
```
Comment (VedantParanjape) [inline wifi_logger.c:206]: In case of UDP, this will return false. But it should return true, Add macro for udp too

something like
```c
#ifdef UDP
ret = true;
#endif
```

File: include/tcp_handler.h (hunk 5 above)
Comment (shivam5594) [reply include/tcp_handler.h:38]: there is no point of using int since error status is saved in errno variable by esp. This is just to say if it was created successfully or not. if not, user can see errno for better handling for their project.

File: include/wifi_logger.h (hunk 6 above)
Comment (shivam5594) [reply include/wifi_logger.h:34]: is_connected is just for checking if connected or not since status is saved in errno by esp

File: wifi_logger.c (hunk 7 above)
Comment (shivam5594) [reply wifi_logger.c:206]: not using it in UDP.

File: wifi_logger.c [hunk 8]
```diff
@@ -237,29 +264,53 @@ void wifi_logger()
...
+	while(1) // To keep trying over and over
+	{
+		while(1) // Try creating and connecting to socket
+		{
+			struct tcp_network_data* handle = malloc(sizeof(struct network_data));
+			if (false == tcp_network_manager(handle))
+			{
+				// Slow down while retrying.
+				vTaskDelay(2000 / portTICK_PERIOD_MS);
+				break;
```
Comment (VedantParanjape) [inline wifi_logger.c:280]: You need to free(handle) before breaking out

File: include/udp_handler.h (hunk 4 above)
Comment (shivam5594) [reply include/udp_handler.h:21]: yes we are not using it but if someone add this repo in their project, the compiler will know that this file is C not C++ and will compile it with C compiler instead of C++ which add random numbers at the end of variable name for polymorphism and linker will fail.

File: wifi_logger.c (hunk 8 above)
Comment (shivam5594) [reply wifi_logger.c:280]: already done inside the tcp_network_manager func when it fails, it calls tcp_close_network_manager.

File: wifi_logger.c [hunk 9]
```diff
@@ -237,29 +264,53 @@ void wifi_logger()
 #ifdef CONFIG_TRANSPORT_PROTOCOL_TCP
 void wifi_logger()
```
Comment (VedantParanjape) [inline wifi_logger.c:268]: 1) Something seems wrong with whitespaces, can you use "TAB" for 4 whitespaces everywhere.
2) You need to make sure we are freeing variable handle whenever breaking out

File: tcp_handler.c (hunk 1 above)
Comment (VedantParanjape) [reply tcp_handler.c]: Cool, sounds good

File: websocket_handler.c (hunk 2 above)
Comment (VedantParanjape) [reply websocket_handler.c:73]: Yes

File: wifi_logger.c (hunk 9 above)
Comment (shivam5594) [reply wifi_logger.c:268]: 1. it shows correctly indented on my IDE. 
2. it is already done in tcp_network_manager before returning

File: include/tcp_handler.h (hunk 3 above)
Comment (shivam5594) [reply include/tcp_handler.h:21]: yes we are not using it but if someone add this repo in their project, the compiler will know that this file is C not C++ and will compile it with C compiler instead of C++ which add random numbers at the end of variable name for polymorphism and linker will fail.

File: include/udp_handler.h (hunk 4 above)
Comment (VedantParanjape) [reply include/udp_handler.h:21]: Oh, sorry, I didn't see the commit message. But have you tested this ? Like using this in a .cpp esp-app ?

File: include/tcp_handler.h (hunk 5 above)
Comment (VedantParanjape) [reply include/tcp_handler.h:38]: Since, we need to include <stdbool.h> to use bool in C and lot of API's return int and not bool in C. I suggested this

File: include/udp_handler.h (hunk 4 above)
Comment (shivam5594) [reply include/udp_handler.h:21]: Yes I am working on another project where I used it and yes it is working.

File: wifi_logger.c (hunk 7 above)
Comment (VedantParanjape) [reply wifi_logger.c:206]: Just for completeness

File: wifi_logger.c (hunk 8 above)
Comment (VedantParanjape) [reply wifi_logger.c:280]: My bad ! Just read it

File: include/tcp_handler.h (hunk 5 above)
Comment (shivam5594) [reply include/tcp_handler.h:38]: For ESP you do not need to add any library. I tried this with only C code project and it works fine. Library internally handles it according to C compiler or C++ with same #ifdef __cplusplus check. it works for both of my C and C++ projects.

Comment (VedantParanjape) [reply include/tcp_handler.h:38]: Ohk, then

File: include/udp_handler.h (hunk 4 above)
Comment (VedantParanjape) [reply include/udp_handler.h:21]: Cool.

So you are using this library in some other project ? If it is open source let me know. I'll add it in readme

File: wifi_logger.c (hunk 9 above)
Comment (VedantParanjape) [reply wifi_logger.c:268]: 1) There's something wrong with github, this happens too often.
2) Ohk, just wanted to make sure, as malloc in a while loop appears to be dangerous

File: include/udp_handler.h (hunk 4 above)
Comment (shivam5594) [reply include/udp_handler.h:21]: Yes I am working on a flight controller and wanted to add MAVLINK support or atleast logging over wifi. For now it is just an IMU code and still building. https://github.com/shivam5594/ESP32_MPU9250

File: wifi_logger.c (hunk 7 above)
Comment (shivam5594) [reply wifi_logger.c:206]: done

File: wifi_logger.c (hunk 9 above)
Comment (shivam5594) [reply wifi_logger.c:268]: 1. i changed my indentation to 4 spaces instead of a tab. it should be fine now
2. very true, that's why i took the log_message variable outside. Constantly allocating and freeing memory was waste of time, i thought

Comment (VedantParanjape) [reply wifi_logger.c:268]: Cool, I'll merge the pull request

**Comment by VedantParanjape** [comment]
@shivam5594 I had to resolve some conflicts to, please check the following commits: https://github.com/shivam5594/esp-wifi-logger/commit/acd8d1b881ed28e5ed08c6d66cc217a1ebe52296

If there's any mistake let me know, if there isn't I'll merge the PR

File: include/udp_handler.h (hunk 4 above)
Comment (VedantParanjape) [reply include/udp_handler.h:21]: Nice, it uses a cam board. If you need any contributor let me know ;)

**Comment by shivam5594** [comment]
Looks good

**Comment by VedantParanjape** [comment]
Thanks for your contribution. I have some more enhancements in mind. I'll refer you in that issue

**Comment by shivam5594** [comment]
sure always up for some work @VedantParanjape
